docs/writing_rules/examples.md

```diff
@@ -0,0 +1,232 @@
+
+```yaml
+patterns:
+- pattern-not: subprocess.$FUNC("...")
+- pattern: subprocess.$FUNC(..., shell=True, ...)
+```
+
+```python
+import subprocess
+import sys
+
+subprocess.call("echo 'hello'") # Doesn't match
+subprocess.call("grep -R {} .".format(sys.argv[1])) # Doesn't match
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True) # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True, cwd="/home/user") # Matches here
+subprocess.run("grep -R {} .".format(sys.argv[1]), shell=True) # Matches here too!
+```
+
+https://semgrep.live/nJ9d
+
+## Enforce Specific Use of an API
+
+Sometimes you may wish to enforce the specific use of a function's API. There are many examples of this, such as `subprocess.call(..., shell=True, ...)` above; you may wish to match and fail any commit where `shell=True`. This is easy to do in Semgrep, as seen in the above section.
+
+However, there are also function APIs that are insecure by default--or insecure depending on context, such as Jinja2, which does [not enable autoescaping by default](https://github.com/pallets/jinja/blob/2a8515d2e53a2be475d9df3fe44e308501201a95/src/jinja2/environment.py#L296). Jinja2 is an arbitrary templating engine, so this makes sense in non-web contexts. This may not be obvious, though, and if you are working directly with the Jinja2 engine in a web context you want to make sure `autoescape=True`.
+
+(Not to scare anyone: Flask, for instance, [autoescapes templates with the '.html' extension](https://github.com/pallets/jinja/blob/2a8515d2e53a2be475d9df3fe44e308501201a95/src/jinja2/environment.py#L296).)
+
+This is an interesting case because we want to enforce the **presence** of `autoescape=True`. *Matching* this is easy:
+
+```yaml
+patterns:
+- pattern: jinja2.Environment(..., autoescape=True, ...)
+```
+
+But what we *want* is to alert when the *opposite* conditions are met. Therefore we want to match (1) when `autoescape=False` **and** (2) when `autoescape` is not present in the function call at all! This pattern will match when `jinja2.Environment()` does not contain `autoescape=True`:
+
+```yaml
+patterns:
+- pattern-not: jinja2.Environment(..., autoescape=True, ...)
+- pattern: jinja2.Environment(...)
+```
+
+🚧🚧 Coming soon: semgrep.live link 🚧🚧
+
+This can be generalized with the following approach:
+
+1. Match the function call by name.
+2. Filter out good patterns.
+
+### Secure Cookies in Flask
+
+Another example of this approach is [setting secure cookies in Flask](https://semgrep.live/EwB5).
```

Comment:
Since the writing order used in above examples includes the semgrep link below the code example, should we replace `[setting secure cookies in Flask]` with a link to the Flask docs, and then include `https://semgrep.live/EwB5` below the patterns?